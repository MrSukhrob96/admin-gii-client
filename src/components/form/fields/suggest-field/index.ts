export * from './suggest-field.component';
