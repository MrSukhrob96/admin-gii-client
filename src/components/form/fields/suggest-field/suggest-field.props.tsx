import { MenuItem, OptionProps } from "@blueprintjs/core";
import { ItemRenderer } from "@blueprintjs/select";
import { useEffect, useState } from "react";
import { Control, ControllerProps, useController } from "react-hook-form";

/**
 * <SuggestField /> props
 */
export type SuggestFieldProps = {
  items: OptionProps[];
  name: string;
  helperText?: string;
  label?: string;
  rules?: ControllerProps['rules'];
  control: Control<any>;
}

export const useSuggestFieldProps = (props?: SuggestFieldProps) => {
  const [selectedItem, setSelectedItem] = useState<OptionProps | null>(null);
  const {
    field: { value, onChange },
    fieldState: { error }
  } = useController({
    name: props.name,
    rules: props.rules,
    control: props.control
  });

  const onItemItemSelect = (item: OptionProps) => {
    onChange(item.value);
  };

  const inputValueRenderer = (item: OptionProps) => item.label;

  const itemRenderer: ItemRenderer<OptionProps> = (item: OptionProps, { handleClick, handleFocus, modifiers, index }) => {
    if (!modifiers.matchesPredicate) {
        return null;
    }
    return (
        <MenuItem
            active={modifiers.active}
            disabled={modifiers.disabled}
            key={item.value}
            onClick={handleClick}
            onFocus={handleFocus}
            text={`${index}. ${item.label}`}
        />
    );
};


  useEffect(() => {
    if (props.items?.length <= 0) return
    const item = props.items.find(item => item.value === value);
    setSelectedItem(item);
  }, [props.items, value])

  return {
    onItemItemSelect,
    inputValueRenderer,
    error,
    itemRenderer,
    selectedItem
  };
};

