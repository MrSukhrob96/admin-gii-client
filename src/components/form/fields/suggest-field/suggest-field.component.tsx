import { FormGroup, MenuItem } from '@blueprintjs/core';
import { Suggest } from '@blueprintjs/select';
import { hoc } from '@utils/hoc';
import { useSuggestFieldProps } from './suggest-field.props';

/**
 * <SuggestField />
 */
export const SuggestField = hoc(
  useSuggestFieldProps,
  ({
    items,
    inputValueRenderer,
    onItemItemSelect,
    name,
    label,
    rules,
    helperText,
    itemRenderer,
    error,
    selectedItem
  }) => (
    <FormGroup
      helperText={error?.message || helperText}
      label={label}
      labelFor={`${name}-input`}
      labelInfo={rules?.required && '(majburiy)'}
      intent={error ? 'danger' : 'none'}
    >
      <Suggest
        items={items}
        inputValueRenderer={inputValueRenderer}
        onItemSelect={onItemItemSelect}
        itemRenderer={itemRenderer}
        selectedItem={selectedItem}
        noResults={<MenuItem disabled={true} text='No results.' />}
        popoverProps={{minimal: true, fill: true, popoverClassName: 'popover-height'}}
        resetOnClose
        resetOnQuery
        fill
      />
    </FormGroup>
  )
);
