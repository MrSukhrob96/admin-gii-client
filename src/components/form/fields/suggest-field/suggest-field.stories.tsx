import { ComponentStory, ComponentMeta } from '@storybook/react';
import { SuggestField } from './suggest-field.component';
import '@styles/index.css';

export default {
  title: 'Common/SuggestField',
  component: SuggestField,
} as ComponentMeta<typeof SuggestField>

const Template: ComponentStory<typeof SuggestField> = (args) => <SuggestField {...args} />;

export const SuggestFieldInitial = Template.bind({});
SuggestFieldInitial.args = {
}