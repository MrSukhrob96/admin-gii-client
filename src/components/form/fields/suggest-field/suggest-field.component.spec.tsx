import { customRender } from '@test-utils';
import { renderHook } from '@testing-library/react-hooks';
import { useForm } from 'react-hook-form';
import { SuggestField } from './suggest-field.component';

describe('SuggestField', () => {
  test('Should render SuggestField component', async () => {
    const { result } = renderHook(() => useForm({
      defaultValues: { test: '' }
    }));
    const { asFragment } = customRender(
      <SuggestField name='test' control={result.current.control} items={[]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});