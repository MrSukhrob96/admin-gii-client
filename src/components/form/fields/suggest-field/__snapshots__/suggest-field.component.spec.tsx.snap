// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestField Should render SuggestField component 1`] = `
<DocumentFragment>
  <div
    class="bp4-form-group"
  >
    <div
      class="bp4-form-content"
    >
      <div
        class="bp4-popover-wrapper bp4-fill"
      >
        <div
          aria-haspopup="true"
          class="bp4-popover-target"
        >
          <div
            class="bp4-input-group bp4-fill"
          >
            <input
              autocomplete="off"
              class="bp4-input"
              placeholder="Search..."
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
